File: biobakery_workflows/scripts/filter_and_trim.R
#!/usr/bin/env Rscript

# load packages
library(dada2); packageVersion("dada2")
#library(ggplot2)
#library(msa)
#library(gridExtra)
#library(phangorn)
library(tools)

## Collect arguments
args <- commandArgs(TRUE)

## Parse arguments (we expect the form --arg=value)
parseArgs <- function(x) strsplit(sub("^--", "", x), "=")
args.df <- as.data.frame(do.call("rbind", parseArgs(args)))

args.list <- as.list(as.character(args.df$V2))
names(args.list) <- args.df$V1

## Arg1 default
if(is.null(args.list$input_dir)) {
  stop("At least one argument must be supplied (input folder).\n", call.=FALSE)
}

# Print args list to STDOUT
for( i in names(args.list) ) {
  cat( i, "\t", args.list[[i]], "\n")
}

# print contents of folder
cat( grep( "*\\.fastq", list.files(args.list$input_dir), value=T ), sep = "\n" )

#print args.list$input_dir
#print args.list$output_dir 

# these variables are passed to the workflow
input.path <- normalizePath( args.list$input_dir )
#input.path <- args.list$input_dir

output.dir <- ifelse( is.null(args.list$output_dir), "output", args.list$output_dir )
pool.samples <- args.list$pool


# Variable "input.path" containing path to input fastq files directory 
# is inherited from wrapper script dada2_cli.r.

input.file.list <- grep( "*fastq", list.files( input.path ), value = T )
#input.path <- normalizePath("input/")

# List of input files

# Sort ensures forward/reverse reads are in same order
fnFs <- sort(grep( "_R1.*\\.fastq", list.files(input.path), value = T ) )
fnRs <- sort(grep( "_R2.*\\.fastq", list.files(input.path), value = T ) )

# Extract sample names, allowing variable filenames; e.g. *_R1[_001].fastq[.gz]

#sample.names <- gsub( "_R1.*\\.fastq(\\.gz)?", "", fnFs, perl = T)
#sample.namesR <- gsub( "_R2.*\\.fastq(\\.gz)?", "", fnRs, perl = T)

sample.names <- gsub( "_R1.*\\.fastq*", "", fnFs, perl = T)
sample.namesR <- gsub( "_R2.*\\.fastq*", "", fnRs, perl = T)

sample.ext <- file_ext(fnFs)
if(identical("gz",sample.ext[1])) sample.ext <- "fastq.gz"

if(!identical(sample.names, sample.namesR)) stop("Forward and reverse files do not match.")

# Specify the full path to the fnFs and fnRs

fnFs <- file.path(input.path, fnFs)
fnRs <- file.path(input.path, fnRs)
cat(fnFs)
cat(fnRs)

# Create filtered_input/ subdirectory for storing filtered fastq reads
filt_path <- file.path(output.dir, "filtered_input") 
ifelse(!dir.exists(filt_path), dir.create(filt_path, recursive = TRUE), FALSE)

readQC.folder <- file.path(output.dir, "Read_QC")
ifelse(!dir.exists(readQC.folder), dir.create(readQC.folder, recursive = TRUE), FALSE)

# Define filenames for filtered input files
#filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
#filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.fastq.gz"))

filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.", sample.ext))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.", sample.ext))


# Filter the forward and reverse reads:
# Note that:
# 1. Reads are both truncated and then filtered using the maxEE expected errors algorighm from UPARSE.
# 2. Reverse reads are truncated to shorter lengths than forward since they are much lower quality.
# 3. _Both_ reads must pass for the read pair to be output.
# 4. Output files are compressed by default.

rd.counts <- as.data.frame(
  filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,200),
                maxN=0, maxEE=c(1,2), truncQ=2, rm.phix=TRUE,
                compress=TRUE, multithread=TRUE) 
)
# Table of before/after read counts
rd.counts$ratio <- round( rd.counts$reads.out / rd.counts$reads.in, digits = 2 )
rd.counts

# Write rd.counts table to file in readQC.folder
saveRDS(rd.counts, paste0( readQC.folder, "/Read_counts_filt.rds" ))
write.table( rd.counts, paste0( readQC.folder, "/Read_counts_after_filtering.tsv" ), sep = "\t", quote = F, eol = "\n", col.names = NA )

File: biobakery_workflows/scripts/filter_and_trim.py
#!/usr/bin/env python

from rpy2 import robjects
from rpy2.robjects.packages import importr
import os.path
import sys
import re
import argparse

base = importr('base')
dada2 = importr('dada2')
gridExtra = importr('gridExtra')
grdevices = importr('grDevices')
ggplot2 = importr('ggplot2')


def parse_arguments(args):
    """ Parse the arguments"""

    parser = argparse.ArgumentParser(
        description="Filter and trim arguments\n",
        formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument(
        "--input",
        help="input directory where fastq(.gz) files are \n[REQUIRED]",
    )
    parser.add_argument(
        "--output",
        help="output directory \n[REQUIRED]",
    )
    parser.add_argument(
        "--maxee",
        help="maxee \n[REQUIRED]",
    )
    parser.add_argument(
        "--trunc_len_max",
        help="truncate at max length\n[REQUIRED]",
    )
    parser.add_argument(
        "--readcounts_tsv_path",
        help="file path to write read counts tsv format\n[REQUIRED]",
    )
    parser.add_argument(
        "--readcounts_rds_path",
        help = "file path to write read counts rds file\n[REQUIRED]",
    )
    parser.add_argument(
        "--plotF",
        help="file to plot forward reads quality\n[REQUIRED]",
    )
    parser.add_argument(
        "--plotR",
        help="file to plot referse reads quality\n[REQUIRED]",
    )
    parser.add_argument(
        "--pair_id",
        help="pair_identifier\n[REQUIRED]",
    )
    parser.add_argument(
        "--threads",
        help="number of threads\n[REQUIRED]",
    )

    return parser.parse_args()


def main():

    # parse arguments from the user
    args = parse_arguments(sys.argv)

    pair_id0 = args.pair_id.split("1_")[0]
    pair_id1 = pair_id0 + "1"
    pair_id2 = pair_id0 + "2"

    fnFs = sorted([f for f in os.listdir(args.input) if re.match(r'(.*)'+ pair_id1 + '(.*)', f)])
    fnRs = sorted([f for f in os.listdir(args.input) if re.match(r'(.*)'+ pair_id2 + '(.*)', f)])

    sample_names = [f.split(pair_id1)[0] for f in fnFs]
    sample_namesR = [f.split(pair_id2)[0] for f in fnRs]

    print(sample_names)

    if (sample_names != sample_namesR):
        sys.exit("ERROR: Forward and reverse files do not match.")

    fnFs = [os.path.join(args.input, f) for f in fnFs]
    fnRs = [os.path.join(args.input, f) for f in fnRs]

    filt_path = os.path.join(args.output, "filtered_input")

    # Create filtered_input/ subdirectory for storing filtered fastq reads
    if not os.path.isdir(filt_path):
        try:
            os.makedirs(filt_path)
        except EnvironmentError:
            sys.exit("ERROR: Unable to create filtered_input directory: ")

    fileext = os.path.splitext(fnFs[1])[-1]
    if fileext == ".gz": fileext = ".fastq.gz"
    print(fileext)

    filtFs = [os.path.join(args.output, filt_path, name + "_F_filt" + fileext) for name in sample_names]
    filtRs = [os.path.join(args.output, filt_path, name + "_R_filt" + fileext) for name in sample_names]

    robjects.globalenv["output"] = args.output
    robjects.globalenv["fnFs"] = fnFs
    robjects.globalenv["fnRs"] = fnRs
    robjects.globalenv["maxee"] = args.maxee
    robjects.globalenv["trunc_len_max"] = args.trunc_len_max
    robjects.globalenv["readcounts_tsv_path"] = args.readcounts_tsv_path
    robjects.globalenv["readcounts_rds_path"] = args.readcounts_rds_path
    robjects.globalenv["reads_plotF"] = args.plotF
    robjects.globalenv["reads_plotR"] = args.plotR
    robjects.globalenv["ext"] = fileext
    robjects.globalenv["filtFs"] = filtFs
    robjects.globalenv["filtRs"] = filtRs
    robjects.globalenv["threads"] = args.threads

    robjects.r('''

    output.path <- normalizePath(output)

    fnF <- unlist(fnFs, use.names=FALSE)
    fnR <- unlist(fnRs, use.names=FALSE)
    filtF <- unlist(filtFs, use.names=FALSE)
    filtR <- unlist(filtRs, use.names=FALSE)

    # Generate plots and save to file
    # Forward reads
    fwd.qc.plots.list <- list()
    for( i in 1 : length(fnF)) {
      fwd.qc.plots.list[[i]] <- dada2::plotQualityProfile(fnF[i])
      rm(i)
    }
    # Save to file
    png(reads_plotF)
    gridExtra::marrangeGrob( fwd.qc.plots.list, ncol=2, nrow=3, top = NULL )
    dev.off()
    rm(fwd.qc.plots.list)

    # Reverse reads
    rev.qc.plots.list <- list()
    for( i in 1 : length(fnR)) {
      rev.qc.plots.list[[i]] <- dada2::plotQualityProfile(fnR[i])
      rm(i)
    }
    # Save to file
    png(reads_plotR)
    gridExtra::marrangeGrob( rev.qc.plots.list, ncol=2, nrow=3, top = NULL )
    dev.off()
    rm(rev.qc.plots.list)


    # Filter the forward and reverse reads:
    # Note that:
    # 1. Reads are both truncated and then filtered using the maxEE expected errors algorighm from UPARSE.
    # 2. Reverse reads are truncated to shorter lengths than forward since they are much lower quality.
    # 3. _Both_ reads must pass for the read pair to be output.
    # 4. Output files are compressed by default.
    trunc_len_max2 <- strtoi(trunc_len_max)
    trunc_len_max1 <- trunc_len_max2 + 40
    maxee1 <- strtoi(maxee)
    maxee2 <- maxee1 * 2

    rd.counts <- as.data.frame(
      filterAndTrim(fnF, filtF, fnR, filtR, truncLen=c(trunc_len_max1,trunc_len_max2),
                    maxN=0, maxEE=c(maxee1,maxee2), truncQ=2, rm.phix=TRUE,
                    compress=TRUE, multithread=as.numeric(threads)) 
    )
    
    # Table of before/after read counts
    rd.counts$ratio <- round( rd.counts$reads.out / rd.counts$reads.in, digits = 2 )
    
    # Write rd.counts table to file in output folder
    saveRDS(rd.counts, readcounts_rds_path )
    write.table( rd.counts, readcounts_tsv_path, sep = "\t", quote = F, eol = "\n", col.names = NA )
    
    ''')

    rd_counts = robjects.globalenv["rd.counts"]

if __name__ == "__main__":
        main()